File: leetcode78.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int> > res;
        res.push_back(vector<int>());
        sort(nums.begin(),nums.end());
        for(int num:nums)
        {
        	int n = res.size();
        	for (int i = 0; i < n; ++i)
        	{
        		auto list = res[i];
        		list.push_back(num);
        		res.push_back(list);
        	}
        }
        return res;
    }
};

File: leetcode78.java
public class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        res.add(new ArrayList<Integer>());
        Arrays.sort(nums);
        for(int i = 1; i <= nums.length; i++){
            dfs(new ArrayList<Integer>(), nums, i, 0, res);
        }
        return res;
    }
    
    private void dfs(List<Integer> cur, int[] nums,int len, int start, List<List<Integer>> res){
        if(cur.size() >= len){
            // end condition
            res.add(new ArrayList<Integer>(cur));
            return;
        }
        // judgement
        for(int i = start; i < nums.length; i++){
            cur.add(nums[i]);
            // recursive
            dfs(cur, nums, len,i+1, res);
            // return previous condition
            cur.remove(cur.size() - 1);
        }
    }
}
